SaraAlert/SaraAlert
# SARAALERT-1436: Asymptomatic case enrollment/workflow updates

# Description
Jira Ticket: [SARAALERT-1436](https://tracker.codev.mitre.org/browse/SARAALERT-1436)

Requirements: [here](https://mitre.sharepoint.com/sites/CoVDT/Shared%20Documents/General/Product%20Owner%20Materials/Requirements/Sara%20Alert%20562-%20Asymptomatic%20Enrolllment%20Updates.pptx?web=1), update from Kelly: [rename "No Reported Symptoms" to "Asymptomatic"
](https://mitrecorp.slack.com/archives/C01A6P94R9B/p1619733024084000?thread_ts=1619136747.050000&cid=C01A6P94R9B)

5/12 meeting summary: enforce hard validation on enrollment for SO date/asymptomatic, but only soft validation when modifying SO date/asymptomatic/lab results/marking as reviewed by warning/prompting user to populate fields required to meet a recovery definition

5/21 update requirements: [here](https://mitre.sharepoint.com/:p:/r/sites/CoVDT/_layouts/15/Doc.aspx?action=edit&sourcedoc=%7BDC57CEC5-492F-4E87-B2ED-E7E628F7DC70%7D)

Tasks/Changes:

- uncomment asymp case enrollment elements and rename to asymptomatic
- add asymptomatic checkbox under symptom onset on patient page for isolation only
  - only SO date or asymptomatic should be populated/checked at a time
  - clearing SO date warns/prompts user to check asymp, unchecking asymp warns/prompts user for SO date
- bring mark as review back for isolation
  - warn/prompt for SO date or asymptomatic when clearing all or only remaining symptomatic assessment
- warn user when removing only remaining positive lab result or specimen collection date of that result
- highlight SO date or asymptomatic on patient page for isolation cases when ineligible for recovery definition
- advanced filter to find records ineligible for recovery definition
- system calculated symptom onset date uses monitoree's time zone for date conversion

TODO:
- react component tests for symptom onset and laboratory modal 
- system tests for modifying records in isolation and checking for validation


## (Feature) Demo/Screenshots

<img width="1680" alt="Screen Shot 2021-05-03 at 1 03 20 PM" src="https://user-images.githubusercontent.com/9956561/116908952-82929c80-ac11-11eb-9611-b71fe7f4858d.png">

## Important Changes

`app/controllers/patients_controller.rb`
- marking reports as reviewed in isolation will also allow `symptom_onset` and `asymptomatic` to be updated

`app/helpers/patient_query_helper.rb`
- advanced filter query for records ineligible for recovery

`app/javascript/components/enrollment/steps/Exposure.js`
- `asymptomatic` is now a saved field

`app/javascript/components/patient/assessment/AssessmentTable.js`
- records can now be marked as reviewed for isolation cases

`app/javascript/components/patient/assessment/actions/ClearAssessments.js`
- clear assessments has been combined into one component
- add ability to populate `symptom_onset` or `asymptomatic` if no more symptomatic assessments

`app/javascript/components/patient/assessment/actions/LastDateExposure.js`
- this component now only contains the LDE field, LDE is also hidden in isolation now

`app/javascript/components/patient/assessment/actions/MonitoringPeriod.js`
- `LastDateExposure` component has been refactored into this component that encompasses all relevant monitoring period fields for exposure and isolation

`app/javascript/components/patient/assessment/actions/SymptomOnset.js`
- `Asymptomatic` checkbox is added below `Symptom Onset` field
- when asymptomatic is unchecked, the modal includes an input to populate `Symptom Onset` similar to LDE/CE logic

`app/javascript/components/patient/laboratory/LaboratoryModal.js`
- add soft warning when case is asymptomatic and the only first positive lab result is edited

`app/models/assessment.rb`
- automatically uncheck `asymptomatic` if symptomatic report exists

# Testing

Edge cases to test:

1) clear user defined SO date when there are no symptomatic reports -> soft warning with option to check asymptomatic
  a) check asymptomatic -> pos lab ? no problem : invalid Asymptomatic
  b) don’t mark as asymptomatic -> invalid SO date
2) edit/delete pos lab for Asymptomatic case -> soft warning, invalid Asymptomatic
3) create/update symptomatic assessment -> Asymptomatic unchecked and disabled, SO updated
4) mark all/only symptomatic report as reviewed for cases with symptomatic reports
  a) user defined SO -> no problem
  b) system defined SO -> warn/prompt user to update SO or Asymptomatic status

# Checklists

**Submitter:**


@holmesie :


@tstrass :


@sgober :
